### ISSUE TITLE
Spire's cli arg handling appears to interfere with user-space cli args

### ISSUE THREAD

**Issue by tomconnors** [epiccastle/spire#54]
Spire's help indicates that spire's options go before the file name: `spire [options] FILE`, but it appears to look at options after the filename as well, which means userspace options must use different names.

``` clojure
(ns provision
  (:require [clojure.tools.cli :as cli]))

(def cli-opts
  [["-h" "--host HOST" "Host namer"]
   ["-i" "--identity IDENTITY" "Identity file"]])

(defn main [args]
  (let [opts (cli/parse-opts args cli-opts)]
    (prn "opts:" opts)))

(main *command-line-args*)
```

``` bash
spire provision.clj -h 123.123.12.12
# ^ gives spire's help

spire provision.clj --host=123.123.12.12
# ^ runs the script, printing the parsed options.
```

**Comment by retrogradeorbit** [comment]
Its standard practice for GNU tools to accept command after the filename (even while help docs list it first). eg.

```shell
$ ls --help | head -1
Usage: ls [OPTION]... [FILE]...
$ ls foo -alF
-rw-r--r-- 1 crispin crispin 0 May 14 16:50 foo
```

It's also a practice to use `--` to mark the end of all flags to the command line processor.

```shell
$ ls foo -- -alF
ls: cannot access '-alF': No such file or directory
foo
```

spire also supports this `--` marker, so:

```shell
$ spire provision.clj -- -h
"opts:" {:options {}, :arguments ["provision.clj" "-h"], :summary "  -h, --host HOST          Host namer\n  -i, --identity IDENTITY  Identity file", :errors nil}
nil
```

If you pass in an `-h` option now, you will however find it doesn't work:

```shell
$ spire provision.clj -- -h foo
"opts:" {:options {}, :arguments ["provision.clj" "-h" "foo"], :summary "  -h, --host HOST          Host namer\n  -i, --identity IDENTITY  Identity file", :errors nil}
nil
```

this is because spire passes all the args unprocessed to your function (in case you want to see what spire was called with) and your cli/parse-opts is _also_ stopping its flag processing at the same `--`.

One way you can deal with this is for your script to manually strip the args proceeding `--` yourself:

```clojure
#!spire --
(ns provision
  (:require [clojure.tools.cli :as cli]))

(def cli-opts
  [["-h" "--host HOST" "Host namer"]
   ["-i" "--identity IDENTITY" "Identity file"]])

(defn main [args]
  (let [opts (cli/parse-opts
              (rest (drop-while #(not= % "--")  args))
              cli-opts)]
    (prn "opts:" opts)))

(main *command-line-args*)
```

Now your `--` call will succeed:

```shell
$ spire provision.clj -- -h foo
"opts:" {:options {:host "foo"}, :arguments [], :summary "  -h, --host HOST          Host namer\n  -i, --identity IDENTITY  Identity file", :errors nil}
nil
```

Also note I put a shebang `#!spire --` with the `--` included, so when the executable bit of the script is set, the following happens:

```shell
$ ./provision.clj -h foo
"opts:" {:options {:host "foo"}, :arguments ["./provision.clj"], :summary "  -h, --host HOST          Host namer\n  -i, --identity IDENTITY  Identity file", :errors nil}
nil
```

You may also want to test for the `--` being present at all before stripping proceeding args. Or you may want to choose args that are mutually exclusive to spire's.

As for changing this behaviour I'm conflicted over the tradeoff as to whether the surprise of someone adding spire flags after the file and being surprised when they don't work (like going back to a failing command and suffixing `--debug-ssh` only to have the same thing run without any debug and no error about the arg not being processed) outweighs the surprise of someone finding their `-h`, `-e`, `-o` or `-v` flags not working without this additional, non-obvious, `--` work.

I am trying to keep the argument footprint of spire as low as possible for this reason. And I think the compromise will be to make it even lower. I would like to keep `-e` as it's very convenient. But I think `-h`, `-o` and `-v` could be removed and those only provided as double dash style arguments.

**Comment by retrogradeorbit** [comment]
fixed in 0.1.0-alpha.10
